webhook/credentials.go

```diff
@@ -0,0 +1,197 @@
+package webhook
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	EnvKeyGoogleApplicationCredentials   = "GOOGLE_APPLICATION_CREDENTIALS"
+	EnvValueGoogleApplicationCredentials = "some-path"
+
+	AnnotationGCPServiceAccount      = "giantswarm.io/gcp-service-account"
+	AnnotationWorkloadIdentityPoolID = "giantswarm.io/gcp-workload-identity-pool-id"
+
+	LabelWorkloadIdentity = "giantswarm.io/gcp-workload-identity"
+
+	VolumeWorkloadIdentityName        = "workload-identity-credentials"
+	VolumeWorkloadIdentityDefaultMode = 420
+	VolumeMountWorkloadIdentityName   = "workload-identity"
+	VolumeMountWorkloadIdentityPath   = "/var/run/secrets/workload-identity"
+
+	TokenExpirationSeconds                   = 7200
+	ServiceAccountTokenPath                  = "token"
+	GoogleApplicationCredentialsJSONPath     = "google-application-credentials.json"
+	ConfigMapKeyGoogleApplicationCredentials = "config"
+)
+
+type CredentialsInjector struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func NewCredentialsInjector(client client.Client, decoder *admission.Decoder) *CredentialsInjector {
+	return &CredentialsInjector{
+		client:  client,
+		decoder: decoder,
+	}
+}
+
+func (w *CredentialsInjector) Handle(ctx context.Context, req admission.Request) admission.Response {
+	logger := w.getLogger(ctx)
+
+	logger.Info("Handling admission request")
+	defer logger.Info("Done")
+
+	pod := &corev1.Pod{}
+	err := w.decoder.Decode(req, pod)
+	if err != nil {
+		logger.Error(err, "no Pod in admission request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if pod.Spec.ServiceAccountName == "" {
+		message := "Pod has no ServiceAccount"
+		logger.Info(message)
+		return admission.Denied(message)
+	}
+
+	serviceAccount, err := w.getServiceAccount(ctx, pod)
+	if k8serrors.IsNotFound(err) {
+		logger.Error(err, "Pod ServiceAccount does not exist")
+		return admission.Errored(http.StatusBadRequest, err)
```

Comment:
I've thought about it, I think the correct thing is to just do `admission.Denied`. I don't think it's a 404 Not Found, since it's the request itself that isn't valid.